zypper.list_pkgs: add parameter for returned attribute selection

moio opened the pull request:
### What does this PR do?

It adds a new optional parameter to `list_pkg` in the `zypper` module to return more data than the version (original reason is that for SUSE Manager integration we also need `arch` and `install_date`). Format is the same of existing method `info_installed`.

### What issues does this PR fix or reference?

None known.

### New Behavior

If the new parameter `attr` is specified, more package attributes are returned.

### Tests written?

Yes - existing tests have been adapted and a new one has been added.

### Other notes

If this PR is accepted we will likely open others to get more attributes, add the same capabilities to the `yumpkg` module and backport this code at least to 2016.11.

This is my first core contribution so please be patient if quality is not yet at the correct level!

@meaksh, @dincamihai can you take a look please?

moio committed:
zypper.list_pkgs: add attr option

This propagates the option to zypper.install as well.

moio committed:
zypper.list_pkgs: add documentation for attr param

moio committed:
zypper_test: correct mock behavior

Previously incorrect, although the test passed

moio committed:
zypper_test: adapt existing unit tests to new code

moio committed:
zypper_test: add test for list_pkgs and attr

salt-jenkins commented:
@moio, thanks for your PR! By analyzing the history of the files in this pull request, we identified @terminalmage, @isbm and @mcalmer to be potential reviewers.

rallytime commented on salt/modules/zypper.py:
```diff
@@ -661,6 +661,16 @@ def list_pkgs(versions_as_list=False, **kwargs):
+
+        Valid attributes are: version, arch, install_time.
+
+            .. versionadded:: 2016.11.5
```
This versionadded should be `Oxygen` instead of `2016.11.5`.

moio replied:
Fixed in 04a1bbbc48ada0c587c67af76944779f0d325863, thanks!

damon-atkins commented:
Here is a windows example, I have been working on.
Windows specific items are prefixed with win, however they could be placed in a sub dictionary.
So something to think about for extra information from other package managers.  I tried to base it off rpm
```
        arch:
            x64
        publisher:
            Fred
        install_date:
            20130304
        provider:
            innosetup  # Windows Packaging Software
        size:
            2475
        summary:
            CPUID HWMonitor 1.21
        version:
            1.21
        win_display_name:
            CPUID HWMonitor 1.21
        win_install_location:
            C:\Program Files\CPUID\HWMonitor\
        win_product_code:
            CPUID HWMonitor_is1
        win_system_component:
            False
        win_uninstall_cmd:
            "C:\Program Files\CPUID\HWMonitor\unins000.exe"
        win_uninstall_quiet_cmd:
            "C:\Program Files\CPUID\HWMonitor\unins000.exe" /SILENT
```

damon-atkins commented on salt/modules/zypper.py:
```diff
@@ -649,10 +649,10 @@ def version_cmp(ver1, ver2, ignore_epoch=False):
 
 
-def list_pkgs(versions_as_list=False, **kwargs):
+def list_pkgs(versions_as_list=False, attr=None, **kwargs):
```
attr=None  should be added to kwargs, so that the list_pkgs is the same on all platforms.
However I suggest you supply all information by default.   versions_as_list=True gives a short version of information any way.

moio replied:
> attr=None should be added to kwargs, so that the list_pkgs is the same on all platforms

While it is true that many modules share the `(versions_as_list=False, **kwargs)` signature, that's definitely not universal:

```
salt$ ack -A3 "def list_pkgs"
[...]

salt/modules/yumpkg.py
585:def list_pkgs(versions_as_list=False, **kwargs):
[...]

salt/modules/dpkg.py
134:def list_pkgs(*packages):
[...]

salt/modules/nix.py
183:def list_pkgs(installed=True,
184-              attributes=True):
[...]

salt/modules/aptpkg.py
1318:def list_pkgs(versions_as_list=False,
1319-              removed=False,
1320-              purge_desired=False,
1321-              **kwargs):  # pylint: disable=W0613
[...]

salt/modules/rpm.py
129:def list_pkgs(*packages):
salt/modules/pkgng.py
[...]

salt/modules/pkgng.py
347:def list_pkgs(versions_as_list=False,
348-              jail=None,
349-              chroot=None,
350-              root=None,
351-              with_origin=False,
352-              **kwargs):
[...]

salt/modules/freebsdpkg.py
255:def list_pkgs(versions_as_list=False, with_origin=False, **kwargs):
[...]
```

That's even more evident when looking at other methods, like `install`:
```
salt$ ack -A15 "def install\("
[...]

salt/modules/npm.py
63:def install(pkg=None,
64-            pkgs=None,
65-            dir=None,
66-            runas=None,
67-            registry=None,
68-            env=None,
69-            dry_run=False,
70-            silent=True):
[...]

salt/modules/pacman.py
438:def install(name=None,
439-            refresh=False,
440-            sysupgrade=False,
441-            pkgs=None,
442-            sources=None,
443-            **kwargs):
[...]

salt/modules/pkgng.py
658:def install(name=None,
659-            fromrepo=None,
660-            pkgs=None,
661-            sources=None,
662-            jail=None,
663-            chroot=None,
664-            root=None,
665-            orphan=False,
666-            force=False,
667-            glob=False,
668-            local=False,
669-            dryrun=False,
670-            quiet=False,
671-            reinstall_requires=False,
672-            regex=False,
673-            pcre=False,
[...]

```

So I got the impression that uniform method signatures was not a design choice here - I personally think making parameters more visible in platform-specific modules has more value. I can of course change this if requested by the Salt team.

moio replied:
> I suggest you supply all information by default.

I disagree here.

Reason is that the returned data has a different structure when the `attr` parameter is set, so any third party application or script must be ready to accept it.

That means that putting the `attr` parameter as a default will make `zypper` return something different from all other backends that must be treated in a special way by default, and will break backwards compatibility. This is not really the intention here - what I would like to see is an option to return more data only if the requester needs it, only if it is prepared to handle it.

isbm replied:
@damon-atkins the `attr=None` is what `lowpkg.info attr=...` does. Nothing to do here with the "standard signatures".

cachedout replied:
I agree with @isbm and @moio here but I'll give @damon-atkins a chance to reply before we move forward with this. Thanks.

damon-atkins replied:
nix.py is new in development
```
    :param bool attributes:
        show the attributes of the packages when listing all packages.
        Default: True
```

Which standard is going to be used attributes=True/False  or attrib=[list of attributes]? If its attrib=[list of attributes] then I suggest nix.py should be changed.

At the end of the day their needs to be common parameters which perform the same thing.  Making salt easier to learn and (mostly) consistent across platforms.

Should be able to sit at the command prompt salt * pkg.list_pkgs  param=value
And platforms which do not support param ignore it and platforms which support param do all the same thing.  The attributes return should have the same meaning across platforms. 

Unfortunately and overall/platform wide  doco for pkg.list_pkgs for example does not exist. So it means developers need to look at all of pkg modules to try and determine consistence approach before coding.

Something like `attrib` (common) and `attrib_all`  this would be attributes which are not common across platforms, like the win_*.  Should platform specific attributes be prefixed?

This topic comes up now and then with regards to pkg modules api. Last time, a Salt Team member change it to (x=default,y=default) and then after discussion, the Salt Team changed it back to using kwargs so its more compatible with other platforms.

Configuration Management Tools should mostly remove the platform specific differences so their is common approach across platforms.

moio replied:
> nix.py is new in development
...
> Which standard is going to be used attributes=True/False or attrib=[list of attributes]? 

This seems to me like a valid but different concern than the two discussed above (what the default behaviour should be and whether the new parameter should be in `**kwargs`). I am now restricting the discussion to this last topic.

The choice of the `attr` name and its format here was taken mainly to stay coherent with the existing [info_installed](https://docs.saltstack.com/en/latest/ref/modules/all/salt.modules.zypper.html#salt.modules.zypper.info_installed) - lacking a broader overview I am open to change that if the Salt team indicates to do so.

isbm replied:
@damon-atkins since the topic is not about "standard signature" anymore, I see very little difference here:

```python
>>> def foo(a=None, **kw): print a, kw
>>> foo(**{'a':'Foo', 'b':'bar'})
Foo {'b': 'bar'}
>>> foo(a='Foo', b='bar')
Foo {'b': 'bar'}
>>> foo(b='bar')
None {'b': 'bar'}
>>>
>>> def foo(**kw): print 'a' in kw and kw.pop('a') or None, kw
>>> foo(a='Foo', b='bar')
Foo {'b': 'bar'}
>>> foo(**{'a':'Foo', 'b':'bar'})
Foo {'b': 'bar'}
>>> foo(b='bar')
None {'b': 'bar'}
```

The only why one would have `attr` inside the `kwargs` is that the code itself takes care of it, or just passed elsewhere, like in case of `info_installed`. I see a very little merit discussing all that.

moio committed:
zypper.list_pkgs: add install_time_date_t as a supported attribute

damon-atkins commented:
People come along add a bit here and their, and list of params builds up. Then you have a set of functions for each platform that have params in different order. kwargs forces the use of param=value and removes issues associated with order of parms. PY3 has an alternative which is everything after ** must be param=value, so in the long term it will not be an issue once PY2 support is dropped.

Salt has a file per platform. Some other CM tools have one file for all platforms and the single functional interface supports all platforms.  

Salt allows  salt * pkg.list_pkgs parm1 parm2 parm3 to be run across many platforms, its one of its best features, but its not consistent across platforms.  It can be used in scripts.

Joe writes a script with `salt * pkg.list_upgrades   true true` every platform will given an error except for pacman and aptpkg

Joe writes a script with `salt * pkg.upgrade True` 95% of  platforms it would set refresh to True, on pkgng & yumpkg it would try to upgrade the software True.  

**What I am arguing for is consistency of input parameters and output.** For example I prefixed the windows only attributes with win_ because I feel they have no other meaning on any other platform.

https://docs.saltstack.com/en/develop/ref/modules/all/salt.modules.pkg.html currently provides index only, it should provide a common list of standard params for each function and the purpose. For example the person who did nix.py attributes would add it to  salt.modules.pkg.html , next person who wants to add attributes to yumpkg would see this and add the same parameter for the same purpose.

Unfortunately kwargs is the only way to get consistency in PY2, other than adding every param to every function for every platform. e.g. chroot, orphan, force, local, dryrun, etc in the same order,

moio commented:
@damon-atkins I think I understand your high-level reasoning here but now I am a bit confused about what you are asking to see changed specifically in this PR.

Do I understand correctly that all you are asking me is to move `attr` to `kwargs`, and then you will approve this PR?

Thanks!

damon-atkins commented:
Two Parts
a) Attributes - Approach needs to work across 90% of pkg implementations.   In some respects I like the  approach if setting it to true and getting everything. **In some respect param could be set to `True` or `all` for everything and a list for restricting the attributes returned.**   I would just output everything and get the salt command to output JSON so I can feed it into something else to filter it.  `salt \* pkg.list_pkgs attribute=True --out=json`   |  json_filter.py I think their needs to be some way of just saying I want all attributes.  Nix is not part of pkg so you will be setting the standard.  rpm calls the attributes `tags` which does not work for me, I suppose they could also be called `properties`.  attributes or properties are find maybe check out some other modules to see if they have this sort of option. (e.g.  failhard is a common parameter name, their might be one for this purpose already)
b) The use of kwargs for consistent approach, which is with the Salt Team.

moio commented:
@cachedout what do you think? I am ok in changing b) but I feel that a) is beyond the scope of this PR. Can it be accepted without it?

JFI: I have a very similar patch ready for `yumpkg` and will open a PR as soon as this one is accepted.

isbm commented:
@cachedout as of my opinion here (in case it matters), I'd say a) is absolutely way off this PR. As of b) I don't really care, as named attributes and keywords are essentially the same thing, except that the named attributes can have default values. So I would personally be OK not changing anything here anymore, but that would be just my opinion. However, again, the a) should be another PR with own tests and reviews.

damon-atkins commented:
Their has been effort for pkg.info_installed to return the same information with the same attributes.
```
salt-call pkg.info_installed which
local:
    ----------
    which:
        ----------
        arch:
            armv7hl
        build_date:
            2015-06-19T03:15:10Z
        build_date_time_t:
            1434683710
        build_host:
            arm02-builder05.arm.fedoraproject.org
        description:
            The which command shows the full pathname of a specified program, if
            the specified program is in your PATH.
        group:
            Applications/System
        install_date:
            1970-01-01T00:45:45Z
        install_date_time_t:
            2745
        license:
            GPLv3
        packager:
            Fedora Project
        release:
            12.fc23
        relocations:
            (not relocatable)
        signature:
            RSA/SHA256, Sun Jun 21 10:58:38 2015, Key ID 32474cf834ec9cba
        size:
            76205
        source:
            which-2.20-12.fc23.src.rpm
        summary:
            Displays where a particular program in your path is located
        url:
            http://www.xs4all.nl/~carlo17/which/
        vendor:
            Fedora Project
        version:
            2.20
```
from aptpkg.py
```
for pkg_name, pkg_nfo in __salt__['lowpkg.info'](*names, failhard=failhard).items():
    t_nfo = dict()
    # Translate dpkg-specific keys to a common structure
    for key, value in pkg_nfo.items():
        if key == 'package':
            t_nfo['name'] = value
        elif key == 'origin':
            t_nfo['vendor'] = value
        elif key == 'section':
            t_nfo['group'] = value
        elif key == 'maintainer':
            t_nfo['packager'] = value
        elif key == 'homepage':
            t_nfo['url'] = value
        else:
            t_nfo[key] = value

    ret[pkg_name] = t_nfo

return ret
```

info_installed for zypper is
```
            version, vendor, release, build_date, build_date_time_t, install_date, install_date_time_t,
            build_host, group, source_rpm, arch, epoch, size, license, signature, packager, url,
            summary, description.

    for pkg_name, pkg_nfo in __salt__['lowpkg.info'](*names, **kwargs).items():
        t_nfo = dict()
        # Translate dpkg-specific keys to a common structure
        for key, value in pkg_nfo.items():
            if isinstance(value, six.string_types):
                # Check, if string is encoded in a proper UTF-8
                if six.PY3:
                    value_ = value.encode('UTF-8', 'ignore').decode('UTF-8', 'ignore')
                else:
                    value_ = value.decode('UTF-8', 'ignore').encode('UTF-8', 'ignore')
                if value != value_:
                    value = kwargs.get('errors') and value_ or 'N/A (invalid UTF-8)'
                    log.error('Package {0} has bad UTF-8 code in {1}: {2}'.format(pkg_name, key, value))
            if key == 'source_rpm':
                t_nfo['source'] = value
            else:
                t_nfo[key] = value
        ret[pkg_name] = t_nfo

    return ret
```

isbm commented:
@damon-atkins I have no idea what is your problem about this. SLES:
```
f20.suse.de:
    ----------
    which:
        ----------
        arch:
            x86_64
        build_date:
            2014-09-21T13:09:10Z
        build_date_time_t:
            1411304950
        build_host:
            sheep17
        description:
            The which command shows the full pathname of a specified program, if
            the specified program is in your PATH.
        group:
            System/Base
        install_date:
            2017-07-18T22:02:56Z
        install_date_time_t:
            1500415376
        license:
            GPL-3.0+
        packager:
            https://www.suse.com/
        release:
            3.180
        relocations:
            (not relocatable)
        signature:
            RSA/SHA256, Sun Sep 21 13:09:25 2014, Key ID 70af9e8139db7c82
        size:
            71466
        source:
            which-2.20-3.180.src.rpm
        summary:
            Displays where a particular program in your path is located
        url:
            http://carlo17.home.xs4all.nl/which/
        vendor:
            SUSE LLC <https://www.suse.com/>
        version:
            2.20
```
Other than your install date/time is broken (or you are actually talking from the past using time machine?!!), and some data in the fields are different (e.g. license name), I see no much issues.

The Dpkg/Apt is problematic one, because they are very different than the RPM. They even do not contain the information when the package was built, only when it was installed. And yeah, I think I have to add translation from `architecture` to `arch` — I thought I did (or was it deleted over time?)...

For the record, `git blame` on those places should give you a hint why I know that. :wink:

moio committed:
zypper.list_pkgs: move attr parameter to kwargs

moio committed:
zypper.list_pkgs: add all option for attr

moio commented:
@cachedout: OK, I have rebased once again and added two commits on top to address the two concerns to @damon-atkins.

Up to you deciding whether to merge them or leave them out - they are small, completely optional and self-contained.

cachedout commented:
I'm happy with this. Nice work @moio.
